### DISCUSSION TITLE
Automating the Extraction of dashboard.json in Terraform

### DISCUSSION THREAD

**Discussion by ebattat** [grafana/grafonnet#97] [Q&A]
The problem:

$ make > dashboard.json
Generate the following output:

```
data.jsonnet_file.dashboard: Reading...
data.jsonnet_file.dashboard: Read complete after 4s [id=fff343c4f694714716c088faa4aff66dbc2f1e763f80e149f87fdd85451cc933]
No changes. Your infrastructure matches the configuration.
Terraform has compared your real infrastructure against your configuration
and found no differences, so no changes are needed.
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
Outputs:
dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   ....
   "title": "Faro dashboard"
}

EOT
```

To obtain a clean dashboard JSON, we execute the following code:

```
with open(sys.argv[1], 'r') as f:
    match = re.search(r'dashboard = <<EOT(.*?)EOT', f.read(), re.DOTALL)
    if match:
        extracted_content = match.group(1).strip()
```
This method may not be optimal for extracting dashboard.json, as parsing could fail if the output contains 'EOT'.

Do you have a safer method to automate the extraction of dashboard.json in Terraform?

**Comment by malcolmholmes** [comment]
You need to show us your jsonnet.

But rather than saying:

```
{
  dashboard: g.dashboard.new(...)
}
```

Just say:

```
dashboard.new(...)
```

**Comment by ebattat** [comment]
its not working with [existing example](https://github.com/grafana/grafonnet/blob/main/examples/terraform/jsonnet/main.libsonnet)

**Comment by Duologic** [comment]
The existing example works quite well for me, can you describe the steps to reproduce with the existing example?

**Comment by ebattat** [comment]
I run the following steps:
1. Update Makefile with ' -auto-approve'
```
.PHONY: apply
apply:
	@cd jsonnet/ && \
		jb install && \
		cd ../ && \
		terraform apply -auto-approve
```
2. terraform init
3. make > dashboard.json
When opening dashboard.json:

```
You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   "graphTooltip": 1,
   "panels": [
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "reqps"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 24
         },
         "id": 1,
         "targets": [
            {
               "datasource": {
                  "type": "prometheus",
                  "uid": "ops-cortex"
               },
               "expr": "sum by (status_code) (rate(request_duration_seconds_count{job=~\".*/faro-api\"}[$__rate_interval]))"
            }
         ],
         "title": "Requests / sec",
         "type": "timeseries"
      }
   ],
   "schemaVersion": 36,
   "time": {
      "from": "now-6h",
      "to": "now"
   },
   "timezone": "utc",
   "title": "Faro dashboard"
}

EOT
```

**Comment by Duologic** [comment]
Ah, I think I understand you now, you are using Terraform to render the jsonnet and then attempt to read the rendered JSON from the Terraform output. 

Perhaps you can just execute the jsonnet without going through Terraform? Or you could use the [`local_file` resource](https://registry.terraform.io/providers/hashicorp/local/latest/docs/resources/file) in Terraform to write files?

**Comment by ebattat** [comment]
Yes, the following working to me:
`jsonnet -J vendor main.libsonnet > /app/dashboard.json`

But I think to use Terraform is recommended way because its auto ident the output json.

**Reply by Duologic** [reply]
I get clear indentation with `jsonnet` too, not sure what you mean.

**Comment by ebattat** [comment]
Did you try to run your terraform example ?

**Reply by Duologic** [reply]
Of course.

I'm not sure what problems you are facing.

**Reply by ebattat** [reply]
When running your example I got the following output:

Its good enough for manual work but problematic for automation:

```
You can apply this plan to save these new output values to the Terraform
state, without changing any real infrastructure.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   "graphTooltip": 1,
   "panels": [
      {
         "datasource": {
            "type": "datasource",
            "uid": "-- Mixed --"
         },
         "fieldConfig": {
            "defaults": {
               "unit": "reqps"
            }
         },
         "gridPos": {
            "h": 8,
            "w": 24
         },
         "id": 1,
         "targets": [
            {
               "datasource": {
                  "type": "prometheus",
                  "uid": "ops-cortex"
               },
               "expr": "sum by (status_code) (rate(request_duration_seconds_count{job=~\".*/faro-api\"}[$__rate_interval]))"
            }
         ],
         "title": "Requests / sec",
         "type": "timeseries"
      }
   ],
   "schemaVersion": 36,
   "time": {
      "from": "now-6h",
      "to": "now"
   },
   "timezone": "utc",
   "title": "Faro dashboard"
}

EOT
```

**Reply by ebattat** [reply]
@Duologic, can you please approve that you get the same results ?

**Reply by Duologic** [reply]
Yes, I get the same results, what's the problem with that?

**Reply by Duologic** [reply]
I can't really change how Terraform outputs work, that is out of scope for this project.

**Reply by malcolmholmes** [reply]
Oh, I think i get it. Instead of using terraform's `output` functionality, you are supposed to use the generated json in the Grafana Terraform provider, in a `grafana_dashboard` resource to deploy it to Grafana. 

As @Duologic suggests, this is working as expected. I'd encourage you to try the Grafana Terraform provider, and experiment more with Terraform.

**Reply by ebattat** [reply]
We plan to build E2E flow to deploy auto Grafana dashboard automatically using GitHub Actions.
Currently the provided output cannot be used automatically because we need to extract the json from it.
```
data.jsonnet_file.dashboard: Reading...
data.jsonnet_file.dashboard: Read complete after 4s [id=fff343c4f694714716c088faa4aff66dbc2f1e763f80e149f87fdd85451cc933]
No changes. Your infrastructure matches the configuration.
Terraform has compared your real infrastructure against your configuration
and found no differences, so no changes are needed.
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
Outputs:
dashboard = <<EOT
{
   "description": "Dashboard for Faro",
   ....
   "title": "Faro dashboard"
}

EOT
```
I thought that there is a way to get only the json from terraform output instead of using the following code:
```
with open(sys.argv[1], 'r') as f:
    match = re.search(r'dashboard = <<EOT(.*?)EOT', f.read(), re.DOTALL)
    if match:
        extracted_content = match.group(1).strip()
```

Is it more clear now ?

**Reply by malcolmholmes** [reply]
If you just want the json, don't use Terraform. Just invoke jsonnet directly. Using Terraform to invoke the jsonnet you are just making life harder for yourself.
